### ISSUE TITLE
SysEx missed end byte (F7) on reception with specific message lengths

### ISSUE THREAD

**Issue by massimofasciano** [lathoub/Arduino-AppleMIDI-Library#57]
Hi,

I'm having a strange issue. I debugged it for a while but I'm not sure which piece of software is to blame.

This is happening on Windows 10 using multiple clients. I tried MIDI-OX with different buffer sizes and also Reaper.
I'm using http://www.tobias-erichsen.de/software/rtpmidi.html to send to my ESP8266.
The bug can be replicated easily by using the SysEx example https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/master/examples/EthernetShield_SysEx/EthernetShield_SysEx.ino. I adapted the init/connect part of the sketch to my ESP8266 but the loop code remains the same.

I noticed that sending the 265 byte sysex block from the example was working without any problems. I received the sysex in MIDI-OX or Reaper with all bytes intact.
When I tried sending it back, I only got 255 bytes ("S" start of sysex segment). The "E" (end of segment) never got there.

Changing the max sysex or packet size on the Arduino side does not make a difference because the sysex blocks appear to be split on the Windows 10 side at 255 bytes max. It's not the client application so I assume that rtpmidi is doing this.

If I send 254 bytes, everything is ok. I get a full sysex message ("F" in the example program).
The funny thing is that if I send 257 bytes, it also works fine. I get a "S" (start) of 255 bytes followed by a "E" (end segment) of 2 bytes. No bytes lost.
In fact, anything which is a multiple of 255 bytes + 1 byte will cause a problem because the last byte "F7" will be sent in a packet by itself and the decoder on the Arduino side will not trigger a SysEx callback for this event.

I don't know who is to blame. Is it allowed for the sender to split the sysex message into smaller packets without inserting any markers ?

I made a few very ugly fixes that make it work on the Arduino side but I'm not sure what the correct fix is.

Basically, we want to treat an orphan F7 as a sysex message and trigger a sysex callback if we are already decoding a sysex message (sort of like a running status).
It appears to work fine if the F7 is not the first byte in the closing packet because it is not picked up as a command and I think the running status code calls the sysex decoder.

Let me know if you need more info.

Regards,
Massimo

**Comment by lathoub** [comment]
Thanks for signalling.
Sounds like your typical boundary use case.

The splitting of the SysEx messages is something I noted when using [WireShark](https://www.wireshark.org) to analyse the SysEx packets (send using MIDIOX over RTPMIDI on Windows). Have you tried to WireShark the content? Should indicate rapidly who is the culprit and who eats the ending 'E' (it could also be the Arduino .ino

Receiving of the SysEx messages is handled mostly in `packet-rtp-midi.h`, look in `decode_sysex_start` and `decode_sysex_end` and then passed up without much modification to the Arduino script.

**Comment by massimofasciano** [comment]
I was not very familiar with the rtp-midi packet format so I looked at the source
https://tools.ietf.org/html/rfc4695#section-3.2

SysEx over rtp-midi follows different rules than "normal" serial SysEx which I'm more familiar with.

```
          -----------------------------------------------------------
         | Sublist Position |  Head Status Octet | Tail Status Octet |
         |-----------------------------------------------------------|
         |    first         |       0xF0         |       0xF0        |
         |-----------------------------------------------------------|
         |    middle        |       0xF7         |       0xF0        |
         |-----------------------------------------------------------|
         |    last          |       0xF7         |       0xF7        |
         |-----------------------------------------------------------|
         |    cancel        |       0xF7         |       0xF4        |
          -----------------------------------------------------------
```

From what I can see here, the bug is in the sender an not in your code. I think the first and middle segments are correctly encoded but the last one contains a single F7 when sending 256 bytes (255 + 1). It should in fact contain 2 x F7.

I will have to confirm this by using Wireshark as you suggested.

**Comment by massimofasciano** [comment]
After a Wireshark test, I can confirm that the problem is with the client (rtpmidi on Windows).

When sending exactly 256 bytes, I get 2 packets on the wire:
1) 255 bytes of data properly encoded (with F0 ... F0)
2) a malformed packet containing a single F7 as payload instead of F7 F7

Any sysex message size that is a multiple of 255 + 1 will cause the problem with the malformed end packet. Any other size is sent without issue.

So no bug in your code. That's good news!

What do you recommend ? If you leave this issue open and mark it as not an internal bug, we can send the link to the rtpmidi author for verification of the issue.

**Comment by lathoub** [comment]
Thanks for diving into this Massimo. I have contacted Tobias Erichsen via email and pointed him to your assessment of the issue (url to this issue). Hopefully he can confirm or point us in another direction.

Do you have a Mac nearby to test the issue? There it should not occur.

**Comment by lathoub** [comment]
Tobias just indicated via email that he found the error (and that you already reported the bug as well). Good!

**Comment by massimofasciano** [comment]
That's great. I'm looking forward to testing the fixed version of rtpMIDI with my ESP8266.
Unfortunately I don't have a mac to test with but I assume it should work.
I have an rtp-midi App on Android but I don't think it can send long SysEx messages. It is a DAW remote controller.

**Comment by lathoub** [comment]
Did you get an update from Tobias?

**Comment by lathoub** [comment]
rtpMidi has had an update (now 1.1.11 (244))
I can find the release notes, so i do not know if this bug is fixed. Can you check?
